Blog.API/Blog.Infrastructure/Services/CategoryService.cs

```diff
@@ -1,36 +1,125 @@
 
 namespace Blog.Infrastructure.Services;
 
-//TODO:Implement methods
 public sealed class CategoryService : ICategoryService
 {
-    public IEnumerable<Category> GetCategories()
+    private readonly IUnitOfWork<Category> _unitOfWork;
+    private readonly IValidator<CreateCategoryRequest> _createCategoryValidator;
+    private readonly IValidator<UpdateCategoryRequest> _updateCategoryValidator;
+    private readonly ILogger _logger;
+    private readonly IMapper _mapper;
+
+    public CategoryService(IUnitOfWork<Category> unitOfWork,
+        IValidator<CreateCategoryRequest> createCategoryValidator,
+        IValidator<UpdateCategoryRequest> updateCategoryValidator,
+        ILogger<CategoryService> logger,
+        IMapper mapper)
     {
-        throw new NotImplementedException();
+        _unitOfWork = unitOfWork;
+        _createCategoryValidator = createCategoryValidator;
+        _updateCategoryValidator = updateCategoryValidator;
+        _logger = logger;
+        _mapper = mapper;
     }
 
-    public Task<Category?> GetCategoryById(Guid id)
+    public async Task<IEnumerable<CategoryResponse>> GetCategories()
     {
-        throw new NotImplementedException();
+        var categories = await _unitOfWork.GenericRepository.Set
+            .Include(c => c.Posts)
+            .ToListAsync();
+
+        foreach (var category in categories)
+        {
+            if (category.Posts != null)
+            {
+                foreach (Post post in category.Posts)
+                {
+                    post.Category = null;
+                }
+            }
+        }
+        return _mapper.Map<IEnumerable<CategoryResponse>>(categories);
     }
 
-    public Task<Category> CreateCategory(Category createCategory)
+    public async Task<CategoryResponse?> GetCategoryById(Guid id)
     {
-        throw new NotImplementedException();
+        Category? category = await _unitOfWork.GenericRepository.Set
+            .Where(c => c.CategoryId == id)
+            .Include(c => c.Posts)
+            .FirstOrDefaultAsync();
+
+        if (category is not null)
+        {
+            if (category.Posts != null)
+            {
+                foreach (Post post in category.Posts)
+                {
+                    post.Category = null;
+                }
+            }
+            return _mapper.Map<CategoryResponse>(category);
+        }
+
+        _logger.LogError("Object with id {id} doesn't exist", id);
+        return null;        
     }
 
-    public Task<Category?> UpdateCategory(Category updateCategory)
+    public async Task<CategoryResponse?> CreateCategory(CreateCategoryRequest createCategory)
     {
-        throw new NotImplementedException();
+        var result = _createCategoryValidator.Validate(createCategory);
+        if (!result.IsValid)
+        {
+            _logger.LogError("CreateCategoryRequest object errors:\n{errors}", result.Errors.Select(e => e.ErrorMessage));
+            return null;
+        }
```

Comment:
I'm not sure if we need to validate requests on service level. I think we should do it on request level